rfcs/328-declarative-configuration.md

```diff
@@ -0,0 +1,149 @@
+# Declarative Configuration
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-goals](#non-goals)
+- [Design Details](#design-details)
+  - [Repository Configuration](#repository-configuration)
+  - [Magefile Targets Implementation Details](#magefile-targets-implementation-details)
+  - [Magefile Targets Available Day One](#magefile-targets-available-day-one)
+  - [Testing Framework](#testing-framework)
+  - [Configuration Examples](#configuration-examples)
+- [Testing and Release Plan](#testing-and-release-plan)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+
+## Summary
+
+This proposal replaces environment variable based configuration being used in Make+Magefiles in favor of using declarative yaml based configuration.
+
+## Motivation
+
+As we've been writing `devbase`, and other tooling that uses it, we've identified that the current practice of environment variable based configuration as well as opinionated and unchangeable defaults is not flexible enough for using this tooling both outside of stencil-base (where it is primarily used today) and giving our users the flexibility to easily make minor behaviour changes. This has, ultimately created two pain points:
+
+ * Documentation is hard to write and not easily discoverable
+ * Finding configuration consumption, and configurable areas is difficult
+ * Using `devbase` outside of `stencil-base` projects is not easy (must use sub-modules, and even then it's hard to "plug and play") 
+
+### Goals
+
+ - All configuration should be written as go structs (as we intend to only write new targets in), and have godoc compatible documentation on each exported field along with examples as needed.
+ - Standard location for all configuration for tooling provided by devbase, et. al
+ - Documentation on how to write Magefile targets
+ - Ensure that all targets are "plug-and-play" compatible (e.g. usable by themselves), they shouldn't require stencil modules to be used, and if they do require certain options to work "out of the box" they should be sane defaults and well documented (as per the above goals)
+ - Not breaking, while we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal
+
+### Non-goals
+
+ - Expose new functionality, this would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.
```

Comment:
Comma splice:
```suggestion
 - Expose new functionality. This would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.
```

Reply:
You've probably realized by now I am terrible about following the rules of comma usage and "to" vs "too"